### PR TITLE
Affix/Overlay > Various enhancements/fixes

### PR DESCRIPTION
**Pull request by Blackbaud-SteveBrush**
I've tested the following changes with the Popover/Dropdown work:

**Affix service changes:**
- Renamed non-public properties from "scrollable" to "overflow" throughout.
- Added `SkyAffixAutoFitContext` to allow the auto-fit functionality to respond to different contexts (either the viewport, or the base element's overflow parent).
- Added `AutoFitOverflowOffset` which will allow the affix service to detect dynamic boundaries (such as the Omnibar, or the dock).
- Added the `offsetChange` and `overflowScroll` events to the `Affixer`.
- Added unit tests to account for various scenarios.

**Overlay service changes:**
- Overlay service destroys the host component if all overlays have been closed.
- Added public `close` method on the service to be used by consumers to close an overlay (instead of using the deprecated `close` method on the overlay's instance).
- Fixed the service to not create multiple `sky-overlay-host` elements.
- Fixed the adapter service to not overwrite `body` styles when restricting body scrolling.
- Fixed overlays getting inserted underneath the Omnibar.
- Fixed the overlay component to not require the `Router` service be injected so that existing consumer unit tests don't need to pull in the `RouterTestingModule` to avoid failures.
- Refactored overlay service unit tests to be easier to manage in the future.

### CODE REVIEW

File: src/app/public/modules/adapter-service/adapter.service.spec.ts [hunk 1]
```diff
@@ -38,7 +38,7 @@ describe('Core adapter service', () => {
   describe('getFocusableChildren', () => {
     it('should return an array of all focusable children', () => {
       const actual = component.getFocusableChildren(nativeElement);
-      expect(actual.length).toEqual(6);
+      expect(actual.length).toEqual(8);
```
Comment (Blackbaud-SteveBrush) [inline src/app/public/modules/adapter-service/adapter.service.spec.ts]: Added some more unit tests to the core adapter service so we could get 100% coverage.

File: src/app/public/modules/affix/affix-auto-fit-context.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+  /**
+   * Auto-fit functionality will respect the nearest overflow parent element's dimensions.
+   */
+  OverflowParent = 0,
+
+  /**
+   * Auto-fit functionality will respect the browser viewport dimensions.
+   */
+  Viewport
```
[the remark this answers is not in the document]
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/affix/affix-auto-fit-context.ts]: TypeScript automatically adds sequential numeric values to each item in the enum (if the first item is listed as `0`), but it is better to be explicit; I'll add it.

File: src/app/demos/affix/affix-demo.component.ts [hunk 3]
```diff
@@ -95,15 +116,15 @@ export class AffixDemoComponent {
...
     const left = baseElement.offsetLeft;
 
-    if (this.enableScrollableParent) {
-      const scrollable: HTMLDivElement = this.parentScrollableRef.nativeElement;
-      scrollable.scrollTop = top -
-        scrollable.offsetTop -
-        (scrollable.clientHeight / 2) +
+    if (this.enableOverflowParent) {
+      const overflowParent: HTMLDivElement = this.parentOverflowRef.nativeElement;
+      overflowParent.scrollTop = top -
```
[the remark this answers is not in the document]
Comment (Blackbaud-SteveBrush) [reply src/app/demos/affix/affix-demo.component.ts]: These calculations are purely for the visual tests to test the various placements and scenarios that the affixed element might be in; I don't anticipate the consumer having to write this level of complexity in their applications.

File: src/app/demos/affix/affix-demo.component.html [hunk 4]
```diff
@@ -3,27 +3,34 @@
     class="canvas"
   >
     <div
-      class="parent-scrollable"
+      class="parent-overflow"
       id="screenshot-affix"
       [ngClass]="{
-        'scroll': enableScrollableParent,
+        'overflow': enableOverflowParent,
```
[the remark this answers is not in the document]
Comment (Blackbaud-SteveBrush) [reply src/app/demos/affix/affix-demo.component.html]: I named the "parent that has an `overflow` setting" the `OverflowParent` throughout the affix source code. The boolean above is enabling/activating an "overflow parent" in order to test the `OverflowParent` context. If that still doesn't sit right with you, we can discuss changing the wording of the context throughout. 💪

File: src/app/demos/affix/affix-demo.component.scss [hunk 5]
```diff
@@ -25,21 +29,21 @@
   visibility: hidden;
 }
 
-.parent-scrollable {
+.parent-overflow {
```
[the remark this answers is not in the document]
Comment (Blackbaud-SteveBrush) [reply src/app/demos/affix/affix-demo.component.scss]: Not sure I understand. The parent containers' dimensions should remain the same size, but the visual test is making sure larger base elements work as expected.

File: src/app/public/modules/affix/affix.directive.ts [hunk 6]
```diff
@@ -81,6 +105,18 @@ export class SkyAffixDirective implements OnChanges, OnDestroy {
...
+   * Fires when the affixed element's offset changes.
+   */
+  @Output()
+  public affixOffsetChange = new EventEmitter<SkyAffixOffsetChange>();
+
+  /**
+   * Fires when the affixed element's overflow container is scrolled.
+   */
+  @Output()
+  public affixOverflowScroll = new EventEmitter<void>();
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/affix/affix.directive.ts]: Should we complete `affixOverflowScroll` and `affixOffsetChange` ?

File: src/app/public/modules/affix/affix.directive.spec.ts [hunk 7]
```diff
@@ -478,11 +559,165 @@ describe('Affix directive', () => {
...
-    componentInstance.enableScrollableParent = true;
+    componentInstance.enableOverflowParent = true;
     componentInstance.scrollTargetToTop();
     fixture.detectChanges();
     fixture.whenStable().then(() => {
       expect(fixture.nativeElement).toBeAccessible();
     });
   }));
+
+  it('should work when the base element is larger than the affixed element', () => {
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/affix/affix.directive.spec.ts]: Would you mind expanding on "it should work..." ? What's the expected result?

File: src/app/public/modules/index.ts [hunk 8]
```diff
@@ -1,11 +1,16 @@
...
+export * from './affix/affix-auto-fit-context';
 export * from './affix/affix-config';
 export * from './affix/affix-horizontal-alignment';
+export * from './affix/affix-offset-change';
+export * from './affix/affix-offset';
 export * from './affix/affix-placement';
 export * from './affix/affix-placement-change';
 export * from './affix/affix-vertical-alignment';
 export * from './affix/affix.module';
+export * from './affix/affix.service';
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/index.ts]: Are consumers supposed to use the service? Or should they only ever be using the directive?

Another (maybe unrelated) question - I noticed the docs page has info for the `createAffixer()` and it links to `SkyAffixer` (but the link is broken). Is that intended to also be exposed to consumers?

File: src/app/demos/affix/affix-demo.component.ts [hunk 9]
```diff
@@ -40,35 +42,54 @@ export class AffixDemoComponent {
...
+  public autoFitOverflowOffset: SkyAffixOffset;
+
   public disabled: boolean = false;
 
   public enableAutoFit: boolean = true;
 
-  public enableScrollableParent: boolean = false;
+  public enableLargeBaseElement: boolean = false;
+
+  public enableFullWidthBaseElement: boolean = false;
```
Comment (Blackbaud-AlexKingman) [inline src/app/demos/affix/affix-demo.component.ts]: Check your alphabetical ordering here.

File: src/app/public/modules/overlay/overlay.service.ts [hunk 10]
```diff
@@ -25,58 +25,79 @@ import {
...
+  public close(instance: SkyOverlayInstance): void {
+    this.destroyOverlay(instance);
+    instance.componentRef.destroy();
+
+    if (SkyOverlayService.overlays.length === 0) {
+      this.removeHostComponent();
+    }
+  }
+
   public closeAll(): void {
```
[the remark this answers is not in the document]
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/overlay/overlay.service.ts]: Good call.

File: src/app/public/modules/affix/affixer.ts [hunk 11]
```diff
@@ -32,13 +40,14 @@ import {
...
   getElementOffset,
-  getScrollableParents,
-  isOffsetVisibleWithinParent
+  getOverflowParents,
+  isOffsetFullyVisibleWithinParent,
+  isOffsetPartiallyVisibleWithinParent
 } from './dom-utils';
 
 const DEFAULT_AFFIX_CONFIG: SkyAffixConfig = {
+  autoFitContext: SkyAffixAutoFitContext.OverflowParent,
```
[the remark this answers is not in the document]
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/affix/affixer.ts]: Unfortunately, no. This is related to another question you asked about referencing default values for complex objects; still working on a solution: https://github.com/blackbaud/skyux-docs-tools/issues/38

File: src/app/public/modules/affix/affixer.ts [hunk 12]
```diff
@@ -117,27 +148,43 @@ export class SkyAffixer {
...
+  }
+
   /**
    * Destroys the affixer.
    */
   public destroy(): void {
     this.reset();
     this._placementChange.complete();
-    this._placementChange = undefined;
+    this._offsetChange.complete();
```
[the remark this answers is not in the document]
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/affix/affixer.ts]: Good catch.

File: src/app/public/modules/affix/affix.directive.ts (hunk 6 above)
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/affix/affix.directive.ts]: Good catch.

File: src/app/public/modules/affix/affix.directive.spec.ts (hunk 7 above)
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/affix/affix.directive.spec.ts]: Yep, that's horribly undescriptive.

File: src/app/public/modules/index.ts (hunk 8 above)
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/index.ts]: Both should be available to consumers (though, they should only be used internal to our components for now). I have an issue to allow docs-tools to document plain old TypeScript classes, here: https://github.com/blackbaud/skyux-docs-tools/issues/30

File: src/app/public/modules/overlay/overlay.component.ts [hunk 13]
```diff
@@ -112,33 +117,44 @@ export class SkyOverlayComponent implements OnInit, OnDestroy {
...
+  public attachTemplate<T>(templateRef: TemplateRef<T>, context: T): EmbeddedViewRef<T> {
+    this.targetRef.clear();
+
+    return this.targetRef.createEmbeddedView(templateRef, context);
   }
 
   private applyConfig(config: SkyOverlayConfig): void {
     this.showBackdrop = config.showBackdrop;
-    this.allowClickThrough = (!this.showBackdrop && !config.enableClose);
+    // this.allowClickThrough = (!this.showBackdrop && !config.enableClose);
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/overlay/overlay.component.ts]: Comments?

Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/overlay/overlay.component.ts]: Gosh darn.

**Pull request merged**